lib/encoding.ts

```diff
@@ -0,0 +1,180 @@
+type CompactableRuleStepKey = 'exists' | 'visible' | 'waitForThenClick' | 'click';
+type CompactNullableBoolean = 0 | 1 | 2;
+
+export type CompactCMPRule = [
+    number, // minimumRuleStepVersion
+    string, // name
+    CompactNullableBoolean, // cosmetic
+    string, // runContext.urlPattern
+    number, // runContext.main + runContext.frame
+    number[], // prehideSelectors
+    CompactCMPRuleStep[], // detectCMP
+    CompactCMPRuleStep[], // detectPopup
+    CompactCMPRuleStep[], // optOut
+    CompactCMPRuleStep[], // test
+    Pick<AutoConsentCMPRule, 'intermediate'>, // extra
+];
+
+export type CompactCMPRuleset = {
+    v: number;
+    s: string[];
+    r: CompactCMPRule[];
+};
+
+/**
+ * Mapping of long to short key for rule step keys that should be shortened, with their value
+ * replaced by an array offset.
+ */
+const compactedRuleSteps: [CompactableRuleStepKey, keyof CompactRulesStep][] = [
+    ['exists', 'e'],
+    ['visible', 'v'],
+    ['waitForThenClick', 'c'],
+    ['click', 'k'],
+];
+
+function encodeNullableBoolean(value: boolean | undefined): CompactNullableBoolean {
+    if (value === true) {
+        return 1;
+    }
+    if (value === false) {
+        return 0;
+    }
+    return 2;
+}
+
+function decodeNullableBoolean(value: CompactNullableBoolean): boolean | undefined {
+    if (value === 1) {
+        return true;
+    }
+    if (value === 0) {
+        return false;
+    }
+    return undefined;
+}
+
+export function encodeRules(rules: AutoConsentCMPRule[]): CompactCMPRuleset {
+    const strings: string[] = [];
+
+    function replaceStrings(selector: string): number {
+        let index = strings.indexOf(selector);
+        if (index === -1) {
+            index = strings.push(selector) - 1;
+        }
+        return index;
+    }
+
+    function encodeRuleStep(step: AutoConsentRuleStep): CompactCMPRuleStep {
+        const clonedStep: CompactCMPRuleStep = { ...step };
+        for (const [longKey, shortKey] of compactedRuleSteps) {
+            if (clonedStep[longKey] && typeof clonedStep[longKey] === 'string') {
+                clonedStep[shortKey] = replaceStrings(clonedStep[longKey]);
+                delete clonedStep[longKey];
+            }
+        }
+        return clonedStep;
+    }
+
+    const compactRules: CompactCMPRule[] = rules.map((r) => {
+        return [
+            r.minimumRuleStepVersion || 0,
+            r.name,
+            encodeNullableBoolean(r.cosmetic),
+            r.runContext?.urlPattern || '',
+            parseInt(`${encodeNullableBoolean(r.runContext?.main)}${encodeNullableBoolean(r.runContext?.frame)}`),
+            (r.prehideSelectors || []).map(replaceStrings),
+            r.detectCmp.map(encodeRuleStep),
+            r.detectPopup.map(encodeRuleStep),
+            r.optOut.map(encodeRuleStep),
+            (r.test || []).map(encodeRuleStep),
+            {
+                intermediate: r.intermediate,
+            },
```

Comment:
Let's make it more explicit?
```suggestion
            r.intermediate !== undefined ? { intermediate: r.intermediate } : {},
```